src/Init/System/Uri.lean

```diff
@@ -0,0 +1,105 @@
+/-
+Copyright (c) 2019 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Lovett
+-/
+
+namespace System
+namespace Uri
+namespace UriEscape
+
+def hexDigitToNat (c : Char) : Option Nat :=
+  if '0' ≤ c ∧ c ≤ '9' then some (c.toNat - '0'.toNat)
+  else if 'a' ≤ c ∧ c ≤ 'f' then some (c.toNat - 'a'.toNat + 10)
+  else if 'A' ≤ c ∧ c ≤ 'F' then some (c.toNat - 'A'.toNat + 10)
+  else none
+
+def decodeUri (uri : String) : String := Id.run do
+  let mut decoded : ByteArray := ByteArray.empty
+  let len := uri.data.length
+  let mut pos : Nat := 0
+  while pos < len do
+    let c := uri.data[pos]!
```

Comment:
Someone should correct me if there's magic to prevent this but I think this may be quadratic since there's an access of `List Char` at increasing indices. `String.get` and `String.next` to advance the `pos` should be linear.

Reply:
rats string[pos] being a linear search sucks, this isn't optimized somewhere deep inside Lean compiler?  This would mean I have to use `for c in uri` and build a proper parsing state machine to remember if I've seen a '%' and so on... yuk.

Reply:
You should use a `String.iterator` (created by `String.mkIterator`), then the code won't fundamentally have to change

Reply:
Well, I guess it would have to change even less using Wojciech's suggestion. You just have to make sure to use `utf8ByteSize` instead of `length`.

Reply:
Ah, I see how to do it now, I switched to String.mkIterator.  Pity the string iterator doesn't have an "eof" property.  If it had that I wouldn't need to track the byte position myself.  Please let me know if I can simplify how I've used the iterator, thanks.

Reply:
Does `Iterator.atEnd` work?

Reply:
Ah, I see, unfortunately atEnd is also quadratic, as it uses utf8ByteSize which is recursively calling `c.utf8Size.toNat` on every char in the string, so it is better to snapshot the utf8ByteSize once outside the loop.

Reply:
You had me worried, but actually it's [constant time](https://github.com/leanprover/lean4/blob/e9545a426f51cdadf447e4e05a677ae846446bf3/src/Init/Prelude.lean#L2301), so we should use it. To be fair Mario just added the docstring, it was hidden in the `extern` before.

Reply:
Ah, I see it's a call out to lean_string_utf8_byte_size.  Well, I can still snapshot this value in a local variable, doesn't hurt, unless you find `while i.2.byteIdx < len do` to be terribly ugly or something... turns out I'm using "i.2.byteIdx" in a few places so I'm now trying to do this, but it doesn't work...

```
while let pos := i.2.byteIdx; pos < len do
    let c := i.curr
    if c == '%' && pos + 1 < len then ...
```
the let pos in the while doesn't make pos available inside the while do block... rats...

Reply:
Ok, I switched to this since it looks nicer:
```lean
  while !i.atEnd do
    let pos := i.2.byteIdx
    ...
```

Reply:
and it seems pretty fast:

```lean
def doIt : IO Unit := do
  let x := fromFileUri? "file://a very long 12000 char Uri string"

def doNTimes (fn : IO Unit) (n : Nat) := do
  let mut i := 0
  while i < n do
    fn
    i := i + 1

#eval timeit "yes" (doNTimes doIt 1000000)   --  431ms
```

I think this means it is about 5 microseconds for the 12k Uri.  Not bad, assuming it is not compiling away the unused variable 'x'.

Reply:
Oh, rats, I checked the main.c file and it was dropping the call to fromFileUri, this version measures the time correctly:

```lean
import Test
import Std.System.Uri

open System.Uri

def doIt (uri: String): IO Nat := do
  let x := fromFileUri? uri
  pure (match x with
  | some y => y.utf8ByteSize
  | none   => 0)

def doNTimes (fn : String → IO Nat) (s: String) (n : Nat) := do
  let mut i := 0
  let mut total := 0
  while i < n do
    let len ← fn s
    total := total + len
    i := i + 1
  pure ()

def main : IO Unit :=
  timeit "Time for 1 thousand calls = " (doNTimes doIt longuri 1000)
```

and it comes out at `Time for 1 thousand calls =  633ms` which is about 0.633ms per call, so this implementation is pretty slow.  I probably should remove the `h2.toString.toUTF8` calls somehow and extract the raw bytes that are already encoded in the string, I can use the byteIdx to figure out how many there are per char....